### ISSUE TITLE
MATLAB within docker and/or singularity 

### ISSUE THREAD

**Issue by ofrei** [comorment/gwas#10]
It would be good to find how to package MATLAB software into a docker and/or singularity container. I think this can be achieved with matlab compiler (https://se.mathworks.com/products/compiler.html ) and matlab runtime ( https://se.mathworks.com/products/compiler/matlab-runtime.html ). This would be easier to explore using some simple matlab program, and a separate docker container - just to test this particular issue of running matlab within docker / singularity.

**Comment by bayramakdeniz** [comment]
It can be done with another container: https://github.com/bayramakdeniz/matlab-octave

**Comment by ofrei** [comment]
Thank you for exploring matlab & octave within singularity!
I hadn't got time to fully test this - after a very quick check I was stuck with this error:
![image](https://user-images.githubusercontent.com/13171449/98831092-67518c00-243b-11eb-8efc-ac0657d67ac5.png)
And for some reason I couldn't get ``chmod 777 magicsquare`` to work within container.

I think a practical case here is to enable a specific MATLAB software within container - i.e. we don't need to provide instructions where users mount their MATLAB software from outside  container. 
Could you include ``magicsquare`` executable within container? I would also simplify it as follows:

- include just ``magicsquare.m`` file  into the container (best to include ``magicsquare.m`` in https://github.com/bayramakdeniz/matlab-octave repository  and use ``ADD magicsquare.m`` within Dockerfile)
- add brief instructions how did you build ``magicsqaure.m`` into the binary  file  ( a step that require full matlab installation - so this is done once outside of the container). I would also include the binary ``magicsquare`` direclty into https://github.com/bayramakdeniz/matlab-octave repository, and also into the container.
- test``singularity exec --no-home octave-matlab.sif magicsquare 5``

**Comment by bayramakdeniz** [comment]
This is interesting, I change the magicsquare.zip now and if you unzip and chmod it, it can be easily run: singularity exec  --bind  magicsquare/for_redistribution_files_only:/execute         octave-matlab.sif         /execute/magicsquare 5

On the other hand,  if we want to run the application placed inside the container, a sandbox container may be required in order to be freed from sudo requirement.

**Comment by ofrei** [comment]
Thanks! I've re-downloaded magicsquare.zip, chmod, and run as fllows:
![image](https://user-images.githubusercontent.com/13171449/99655083-8a8fc300-2a5b-11eb-9e04-0f80acbead46.png)

This might be relevant:
https://se.mathworks.com/matlabcentral/answers/373521-why-do-i-receive-could-not-access-mcr-component-cache-when-running-an-executable-on-the-deploymen

Could you try running the MATLAB container on TSD ?

**Comment by bayramakdeniz** [comment]
This is strange, thanks for the suggestion. I will work on that.

For TSD when I run it without .sh as we done in our local, it works but the problem arises when we are using octave-matlab_slurm.sh

When I run 'sbatch octave-matlab_slurm.sh'  I faced with this error: ERROR  : "Unknown image format/type: container_name.sif"

ABORT  : Retval = 255

which is -at first glance- seems  due to version difference on singularity (v2 and v3) (see: https://stackoverflow.com/questions/59491868/building-v2-compatible-singularity-images-with-the-version-3-of-singularity)  Then I  converted it to .simg via

singularity build octave-matlab.simg octave-matlab.sif

and tried to run with octave-matlab.simg instead of .sif and get the same error.

Then I tried  "hello.sif" container and it works without any problem. Since you have built "hello.sif" , my question is did you build it with singularity 2 or did you do any other trick? or did you face the same problem when you run octave-matlab.sif on TSD?

**Comment by ofrei** [comment]
@bayramakdeniz I built ``hello.sif`` with singularity 2.4.2 on my laptop. I agree this version is too old. Will follow up with TSD to enable recent singularity for SLURM.

**Comment by ofrei** [comment]
@bayramakdeniz I've submitted a ticket, but it may take a while to enable singularity ``3.5`` or ``3.6`` on Colossus (i.e. in SLURM jobs). I suggest for now to test your singularity only "locally" (i.e. by running them on ``p33-appn-norment01``; make sure ``which singularity`` points to locally installed version in ``/usr/bin`` - if not, run ``module purge`` to get rid of previously loaded singularity module). 

I'd like to clarify what in my view would be our  main use case of MATLAB . Our goal is to enable users to run specific software such as https://github.com/precimed/pleiofdr or https://github.com/precimed/mostest . Ideally, we should run them with MATLAB re-distributable (not octave). Ideally,  a binary (pre-compiled) version of these packages should be included within MATLAB container (so that there is no need for users to download separate files) - but if this imply that container need to be ``writable`` (aka ``sandbox``) then it's OK to distribute software in two pieces: a base container with MATLAB redistributable, and a pre-compiled binary (i.e. ``pleiofdr`` and ``mostest``) that is compatible with the base MATLAB container. ``magicsquare`` sounds like a good example - if we can make it work, I think we can also use enable pleiofdr and mostest  packages in a similar way.

Octave support, I think, is not a high priority - I don't have specific software that we want to use through Octave. I would only fall back to Octave as last resort if we can't enable pleioFDR and MOSTest in a container with MATLAB redistributable.

**Comment by bayramakdeniz** [comment]
@ofrei, here are the updates:

1-  The container has been updated to v_1.1. In this version, you can run it as : singularity exec --no-home  --bind  magicsquare/for_redistribution_files_only:/execute         octave-matlab.sif         /execute/magicsquare 5

2- I think the best option is mounting pre-compiled versions of projects to the container. Putting them inside the container can be possible via sandbox or some other tricky options but my opinion is mounting them to the container is better. For example, for the latter option,  if we want to update our Matlab Project, then we need to re-build the whole container. On the other hand, if we use the mounting option we do not need to do that (we just need to re-compile the application and mount it).

3- For octave, let's discuss in more detail when we meet.

**Comment by ofrei** [comment]
@bayramakdeniz Could you double-check if it's still working after my changes? Matlab container is special in a sense that more prerequisites are needed to build it (i.e., Matlab needs to be installed).  Few more things:
* would it make sense to   split it into ``matlab`` and ``octave`` containers? I think that's easier for us to maintain - and I don't see whether it'll make a difference for users
* could you document steps needed to build ``magicsquare`` with matlab compiler? I need to learn how this is done and reproduce it on my machine before we can start including more complex software (particularly https://github.com/precimed/pleiofdr and https://github.com/precimed/mostest ) .

**Comment by bayramakdeniz** [comment]
I have seperated this container as matlabruntime and octave.  For now, they are outside from our main gwas folder: https://github.com/comorment/matlabruntime
https://github.com/comorment/octave

I have tested both local and TSD. I have also tested matlabruntime with a slurm script. All works well. I have also added how to build a standalone application. In fact, it is so easy and can be done with one line command. Could you try them and notify me if you face any issue, please?

**Comment by ofrei** [comment]
I'm closing this ticket - it works to run magic square example with matlabruntime.sif.
However will open another ticket to enable pleioFDR without user having to "mcc -m pleiofdr.m"
